moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,365 @@
+import com.squareup.moshi.internal.jsonAnnotations
+import com.squareup.moshi.internal.toStringWithAnnotations
+import java.io.IOException
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.Method
+import java.lang.reflect.ParameterizedType
+import java.lang.reflect.Type
+
+internal class AdapterMethodsFactory(
+  private val toAdapters: List<AdapterMethod>,
+  private val fromAdapters: List<AdapterMethod>
+) : JsonAdapter.Factory {
+  override fun create(
+    type: Type,
+    annotations: Set<Annotation>,
+    moshi: Moshi
+  ): JsonAdapter<*>? {
+    val toAdapter = get(toAdapters, type, annotations)
+    val fromAdapter = get(fromAdapters, type, annotations)
+    if (toAdapter == null && fromAdapter == null) return null
+
+    val delegate: JsonAdapter<Any>? = if (toAdapter == null || fromAdapter == null) {
+      try {
+        moshi.nextAdapter(this, type, annotations)
+      } catch (e: IllegalArgumentException) {
+        val missingAnnotation = if (toAdapter == null) "@ToJson" else "@FromJson"
+        throw IllegalArgumentException(
+          "No $missingAnnotation adapter for ${type.toStringWithAnnotations(annotations)}",
+          e
+        )
+      }
+    } else {
+      null
+    }
+
+    toAdapter?.bind(moshi, this)
+    fromAdapter?.bind(moshi, this)
+
+    return object : JsonAdapter<Any>() {
+      override fun toJson(writer: JsonWriter, value: Any?) {
+        if (toAdapter == null) {
+          requireNotNull(delegate).toJson(writer, value)
+        } else if (!toAdapter.nullable && value == null) {
+          writer.nullValue()
+        } else {
+          try {
+            toAdapter.toJson(moshi, writer, value)
+          } catch (e: InvocationTargetException) {
+            val cause = e.cause
+            if (cause is IOException) throw cause
+            throw JsonDataException("$cause at ${writer.path}", cause)
+          }
+        }
+      }
+
+      override fun fromJson(reader: JsonReader): Any? {
+        return if (fromAdapter == null) {
+          requireNotNull(delegate).fromJson(reader)
+        } else if (!fromAdapter.nullable && reader.peek() == JsonReader.Token.NULL) {
+          reader.nextNull<Any>()
+          null
+        } else {
+          try {
+            fromAdapter.fromJson(moshi, reader)
+          } catch (e: InvocationTargetException) {
+            val cause = e.cause
+            if (cause is IOException) throw cause
+            throw JsonDataException("$cause at ${reader.path}", cause)
+          }
+        }
+      }
+
+      override fun toString() = "JsonAdapter$annotations($type)"
+    }
+  }
+
+  companion object {
+    fun get(adapter: Any): AdapterMethodsFactory {
+      val toAdapters = mutableListOf<AdapterMethod>()
+      val fromAdapters = mutableListOf<AdapterMethod>()
+
+      var clazz = adapter.javaClass
+      while (clazz != Any::class.java) {
+        clazz.declaredMethods.forEach { declaredMethod ->
```

Comment:
Use a for loop

Reply:
https://github.com/square/moshi/pull/1467/commits/9b44ebfe49dbf2db6e0fb10577a2540450dfaf56